### PR TITLE
refactor(planner): minimize ScalarExpr

### PR DESCRIPTION
**Pull request by andylokandy**
- refactor(planner): minimize ScalarExpr
- improve

I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

## Summary

Summary about this PR

Closes #issue

### CODE REVIEW

**Review by sundy-li — APPROVED** [review]
Nice simplification!

**Review by xudong963 — APPROVED** [review]
After the ticket, I think tpcds q4 stackoverflow will be solved, please check again @RinChanNOWWW

**Comment by BohuTANG** [comment]
Conflicting files
src/query/sql/src/planner/semantic/type_check.rs

**Comment by RinChanNOWWW** [comment]
> After the ticket, I think tpcds q4 stackoverflow will be solved, please check again @RinChanNOWWW

It seems that it will also stackoverflow, because the recursion counts are the same.

**Review by leiysky — APPROVED** [review]

**Comment by andylokandy** [comment]
This PR hit on a weird bug that could be solved by https://github.com/datafuselabs/databend/pull/10314

File: src/query/sql/src/planner/optimizer/heuristic/subquery_rewriter.rs [hunk 1]
```diff
@@ -471,7 +471,7 @@ impl SubqueryRewriter {
                             span: subquery.span,
                             func_name: op.to_func_name().to_string(),
                             params: vec![],
-                            arguments: vec![left_condition, right_condition],
+                            arguments: vec![right_condition, left_condition],
```
Comment (RinChanNOWWW) [inline src/query/sql/src/planner/optimizer/heuristic/subquery_rewriter.rs:474]: Why change the order?

Comment (andylokandy) [reply src/query/sql/src/planner/optimizer/heuristic/subquery_rewriter.rs:474]: It was originally reverted

**Pull request merged**